beamer/agent.py

```diff
@@ -84,37 +96,43 @@ def _init(self) -> None:
 
         self._event_processor = EventProcessor(self.context)
 
-        self._event_monitor_l2a = EventMonitor(
-            web3=w3_l2a,
-            contracts=(request_manager,),
-            deployment_block=l2a_contracts_info["RequestManager"].deployment_block,
-            on_new_events=[self._event_processor.add_events],
-            on_sync_done=self._event_processor.mark_sync_done,
-            poll_period=POLL_PERIOD,
-        )
-
-        self._event_monitor_l2b = EventMonitor(
-            web3=w3_l2b,
-            contracts=(fill_manager,),
-            deployment_block=l2b_contracts_info["FillManager"].deployment_block,
-            on_new_events=[self._event_processor.add_events],
-            on_sync_done=self._event_processor.mark_sync_done,
-            poll_period=POLL_PERIOD,
-        )
+        self._event_monitors: dict[ChainId, EventMonitor] = {}
+
+        event_monitor_data = {
+            "on_new_events": [self._event_processor.add_events],
+            "on_sync_done": self._event_processor.mark_sync_done,
+            "poll_period": POLL_PERIOD,
+        }
+
+        for chain_id in chain_ids:
+            if chain_id == source_chain_id:
+                chain_related_data = {
+                    "web3": w3_source,
+                    "contracts": (source_request_manager, source_fill_manager),
+                    "deployment_block": _get_deployment_block(source_contracts_info),
+                }
+            else:
+                chain_related_data = {
+                    "web3": w3_target,
+                    "contracts": (target_request_manager, target_fill_manager),
+                    "deployment_block": _get_deployment_block(target_contracts_info),
+                }
+            event_monitor_data.update(chain_related_data)
+            self._event_monitors[chain_id] = EventMonitor(**event_monitor_data)  # type: ignore
```

Comment:
I would avoid updating `event_monitor` on every loop iteration:
```python
            event_monitor_data.update(chain_related_data)
```
because even though it currently works, it's relatively fragile to reuse the same `event_monitor_data` dict instance.

Instead, I would suggest something like
```python
            kwargs = dict(event_monitor_data, **chain_related_data)
            self._event_monitors[chain_id] = EventMonitor(**kwargs)  # type: ignore
```

Reply:
Having thought some more about this, I now think we should rewrite the whole function so that there is almost no distinction between source and target chain IDs (the exceptions would be e.g. checvks like the one for max validity period above).

So, instead of having `source_contracts` and `target_contracts` and similarly for other variables, we should intend to treat all chains the same way, to the fullest possible extent.

To that end, I think a dict such as the following could help:

```python
   chains: dict[ChainID, dict[str, Any]] = {}
```

(not sure about the value type here yet, but something along those lines).
This would bring us closer to what we want for the multiple-chain-pairs support.
Of course, for now we would still honor the `source-chain` and `target-chain` config values, but later in the epic we will simply deprecate those and look at everything.

So what I propose to do here would be:

- for each non-L1 chain from the config:
- - if the chain is not one of (source-chain, target-chain), continue [note: this will be dropped later]
- - build the contract info, web3 and all other needed objects and put them into the `chains` dict under the chain_id key
- perform the sanity checks we do now [this is one place where we need to distinguish between source and target chains]
-  create the context for the source and target chain [this is another place where we need to distinguish between them]
- for each chain in `chains`:
- - create an appropriate event monitor instance [note: there should be no need for an `if ` here]

Reply:
can we continue this discussion somewhere? I think this is important. 
We do have and will continue to have the concept of source and target chain in the EventProcessor. Will will just not have it anymore on the agent's level.

Having read the above, I think there are some details missing in the Epic to make the new architecture work with our current setup. The Event processors will be created generically by the config. For each creation there exists a source and a target chain. 
I just realized that we need to have the option to create a A -> A EventProcessor out of the config. So let's move the above to a new issue or to the respective epic, please.

Reply:
> can we continue this discussion somewhere? I think this is important.
> ...

Reading the above, I think we are pretty much on the same page. I would have preferred to have made a step in this direction as part of this PR, but it does not really matter so much, as long as we do it at some point. Sooner or later we will run into the need for this so it is fine.